### PR TITLE
[DS-3006] Add explicit data support

### PR DESCRIPTION
**Pull request by nathancooperjones**
## Describe the big picture - why are we doing this?
Closes https://github.com/ShopRunner/collie_recs/issues/10. 

## Any additional details to clarify code in the PR?
My hope was to introduce explicit data support with 1) no breaking changes to existing pipelines, 2) a minimal amount of API changes from the implicit counterpart, and 3) lots of data quality checks to reduce mistakes. If you see any improvements in here that I should add, please let me know! 

## Pull Request Checklist
 - [x] Pull request includes a description of why we are doing this
 - [x] ``CHANGELOG`` has been updated
 - [x] Version in ``_version.py`` has been updated
 - [x] All tests in the ``tests`` folder pass with a local build
 - [x] ``README`` has been updated (if applicable)
 - [x] Documentation in ``docs`` has been updated (if applicable)
 - [x] ``requirements.txt`` and ``requirements-dev.txt`` have been recompiled (if applicable)

### CODE REVIEW

**Review by peterskipper — CHANGES_REQUESTED** [review]
Looking good! Added testing is explicit and seems to have a lot of new coverage. I left some questions/comments.

High level, we should also have a discussion about how this affects training a weighted mashup model downstream. My current implementation relies on (re)defining the ratings matrix to hold e.g. 1 if the interaction is an order and 2 if it's a pview, and then applying a different weight to the loss. This requires the dataset to be explicit I think, but I still want to be able to use negative samples with the loss function, which I *think* might break the logic here? Maybe I shouldn't use the matrix this way and we could create a different sparse matrix holding interaction type / maybe we should change this refactor to allow for this use case since it's coming up? Let's talk it through and decide on a path forward.

File: collie_recs/cross_validation.py [hunk 1]
```diff
@@ -22,24 +26,33 @@ def _subset_interactions(interactions: Interactions, idxs: Iterable[int]) -> Int
...
+            num_items=interactions.num_items,
+            check_num_negative_samples_is_valid=False,
+            max_number_of_samples_to_consider=interactions.max_number_of_samples_to_consider,
+            seed=interactions.seed,
+        )
+    else:
+        return ExplicitInteractions(
+            mat=coo_mat,
+            allow_missing_ids=True,
+            remove_duplicate_user_item_pairs=False,
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/cross_validation.py:45]: Explicit datasets don't have negative samples, so the `check_num_negative_samples_is_valid` doesn't exist!

File: collie_recs/interactions/datasets.py [hunk 2]
```diff
@@ -13,7 +14,7 @@
 import collie_recs
 
 
-class Interactions(torch.utils.data.Dataset):
+class BaseInteractions(torch.utils.data.Dataset, metaclass=ABCMeta):
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/interactions/datasets.py:17]: It's probably way more complicated, but for me, it just makes sure that you can't just instantiate a `BaseInteractions` object - you'll get an error. It's a great way to make a parent, abstract class.

File: collie_recs/model/base/base_pipeline.py [hunk 3]
```diff
@@ -229,21 +251,42 @@ def _configure_loss(self) -> None:
 
         if callable(self.loss):
             self.loss_function = self.loss
+            return
+
+        # explicit losses first
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/model/base/base_pipeline.py:256]: I'm not sure I fully understand the question, so the answer below might not be super relevant: 

We'll only use one loss per model, either implicit or explicit. I added some data quality checks to ensure that `Interactions` objects are only used with implicit losses and `ExplicitInteractions` are only used with explicit losses. 

We'll run this loss check each time in order without checking interactions type yet, and then do that type in the actual loss calculation. 

Does that answer your question?

File: collie_recs/model/base/base_pipeline.py [hunk 4]
```diff
@@ -468,37 +511,77 @@ def validation_epoch_end(
...
+
+            # explicit data
+            (users, items, ratings) = batch
+
+            users = users.long()
+            items = items.long()
+            ratings = ratings.float()
+
+            # get predictions from model
+            preds = self(users, items)
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/model/base/base_pipeline.py:579]: Yup, you got it!

File: tests/test_cross_validation.py [hunk 5]
```diff
@@ -198,60 +260,68 @@ def test_stratified_split_with_user_with_only_one_interaction(
 
 
 class TestSplitsWithWrongP:
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply tests/test_cross_validation.py:262]: Hanna suggested in a previous PR to group similar tests together into a class: https://github.com/ShopRunner/collie_recs/pull/9#discussion_r620483902

File: collie_recs/metrics.py [hunk 6]
```diff
@@ -367,17 +375,132 @@ def evaluate_in_batches(
     all_scores = [acc_score / len(test_users) for acc_score in accumulators]
 
     if logger is not None:
-        try:
-            step = model.hparams.get('num_epochs_completed')
-        except torch.nn.modules.module.ModuleAttributeError:
-            # if, somehow, there is no ``model.hparams`` attribute, this shouldn't fail
-            step = None
+        _log_metrics(model=model,
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/metrics.py:378]: If `step is None`, the logging will still work, so no need for the `try/except` block here after all!

File: collie_recs/metrics.py [hunk 7]
```diff
@@ -367,17 +375,132 @@ def evaluate_in_batches(
...
+
+    try:
+        metrics_dict = dict(zip([x.__name__ for x in metric_list], all_scores))
+    except AttributeError:
+        metrics_dict = dict(zip([type(x).__name__ for x in metric_list], all_scores))
+
+    if verbose:
+        print(f'Logging metrics {metrics_dict} to ``logger``...')
+
+    logger.log_metrics(metrics=metrics_dict, step=step)
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/metrics.py:506]: Yup!!

File: collie_recs/interactions/datasets.py [hunk 8]
```diff
@@ -117,18 +112,6 @@ def __init__(self,
                         'Length of ``ratings`` must be equal to lengths of ``users``, ``items``.'
                     )
 
-                if 0 in set(ratings):
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/interactions/datasets.py:120]: They're allowed for explicit interactions (for now) but not for implicit interactions. This check was just moved further down in the file (line 293).

[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/interactions/datasets.py:120]: If you think there's a reason to add this check in for explicit data as well, let me know!

File: collie_recs/interactions/datasets.py [hunk 9]
```diff
@@ -161,20 +144,183 @@ def __init__(self,
...
+
+    .. list-table::
+        :header-rows: 1
+
+        * - ``__getitem__`` Format
+          - Expected Meaning
+          - Model Type
+        * - ``((X, Y), Z)``
+          - ``((user IDs, item IDs), negative item IDs)``
+          - **Implicit**
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/interactions/datasets.py:232]: I'm not sure I understand this comment - where should this be moved to in the docstring?

File: collie_recs/interactions/dataloaders.py [hunk 10]
```diff
@@ -67,15 +70,19 @@ def mat(self) -> coo_matrix:
...
-    training speed is not a concern. This is the default ``DataLoader`` for ``Interactions``
-    datasets.
+    A light wrapper around a ``torch.utils.data.DataLoader`` for ``Interactions``-type datasets.
+
+    For implicit data, batches will be created one-point-at-a-time using exact negative sampling
+    (unless configured not to in ``interactions``), which is optimal when datasets are smaller
+    (< 1M+ interactions) and model training speed is not a concern. This is the default
+    ``DataLoader`` for ``Interactions`` datasets.
+
+    For explicit data, negative sampling is not used, but batches will still be created
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/interactions/dataloaders.py:80]: Here's what I'm thinking for the structure / modeling format: 

``((X, Y), Z)`` = ``((user IDs, item IDs), negative item IDs)`` = Implicit Model 
``(X, Y, Z)`` = ``(user IDs, item IDs, ratings)`` = Explicit Model 
``((X, Y), Z, C)`` = ``((user IDs, item IDs), negative item IDs, interaction type)`` = Mash Up Implicit Model 

I think the checking logic for this could be something like, roughly: 

```python
if isinstance(batch, Callable) and not isinstance(batch, torch.tensor):
    if len(batch) == 2:
        # implicit model
    elif len(batch) == 3:
        # mash up implicit
else:
    # explicit
```

How does this sound to you?

[the remark this answers is not in the document]
Comment (peterskipper) [reply collie_recs/interactions/dataloaders.py:80]: Sorry for the **slow** reply. So we agree that the Mashup Implicit model would just repurpose the ratings matrix to hold the interaction type? Or would we create a different object for Mashups with a new name?

In principle this works but I'd be worried downstream that checking the `len(batch)` could cause us problems later in ways we wouldn't predict. Like if we refactor either of the model classes we might not think to check here? If possible, I think it'd be better to explicitly check the batch type or model type against a string name. If that's a huge hassle maybe we could get around it with some explicit unit tests of each data loader?

Lmk what you think.

File: collie_recs/interactions/datasets.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (peterskipper) [reply collie_recs/interactions/datasets.py:120]: no, this is great!

File: collie_recs/model/base/base_pipeline.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (peterskipper) [reply collie_recs/model/base/base_pipeline.py:256]: nm, just got confused! This is just running checks not actually calculating a loss :dissapear:

File: collie_recs/interactions/datasets.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (peterskipper) [reply collie_recs/interactions/datasets.py:232]: Hmm no I think this is my fault. This is like markdown or sthg? I think I just misunderstood the structure. If it looks good in read the docs, I'm fine

**Comment by peterskipper** [comment]
Thanks for all your comments, Nate! I'm understanding this better. I continued our discussion on how to handle implicit mashup modelds. And did you decide to add a warning when I e.g.  pass in hinge instead of adaptive_hinge in the base pipeline (see comment above)?

**Comment by nathancooperjones** [comment]
> Thanks for all your comments, Nate! I'm understanding this better. I continued our discussion on how to handle implicit mashup modelds. And did you decide to add a warning when I e.g. pass in hinge instead of adaptive_hinge in the base pipeline (see comment above)?

Yes I did, see here for that warning: https://github.com/ShopRunner/collie_recs/blob/eba517872438d3b33d96fc46b54be92e8949ad9f/collie_recs/model/base/base_pipeline.py#L283-L290

File: collie_recs/interactions/dataloaders.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/interactions/dataloaders.py:80]: I'm not sure of a better solution for handling this right now. 

I've added a good chunk of warnings that will fail training a model without having to rely on the batch first (see something like this: https://github.com/ShopRunner/collie_recs/blob/eba517872438d3b33d96fc46b54be92e8949ad9f/collie_recs/model/base/base_pipeline.py#L557)

I'm hoping that with these extra checks in place in both the datasets and models, we won't have to only rely on the `len(batch)` method. 

For the mashup model, I think it makes sense to just repurpose the ratings matrix for this purpose. We _might_ want to create a whole separate object for that (unless there really aren't a ton of changes and it works for instances where we have and don't have this data).

[the remark this answers is not in the document]
Comment (peterskipper) [reply collie_recs/interactions/dataloaders.py:80]: OK, sounds good. Do you think we can add a unit test for each data loader to pull in a small dataset, use it to make a batch, and then `assert` that the batch is the correct length and/or contains the correct items? Not sure how difficult that would be. If it's doable, that would at least prevent us from accidentally refactoring in such a way that the dataloader would get confused about the batch.

[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/interactions/dataloaders.py:80]: See the tests here: 

https://github.com/ShopRunner/collie_recs/blob/6257f1b05e520c0c5352ccba875d995aeb74c40c/tests/test_interactions.py#L643

https://github.com/ShopRunner/collie_recs/blob/6257f1b05e520c0c5352ccba875d995aeb74c40c/tests/test_interactions.py#L689

https://github.com/ShopRunner/collie_recs/blob/6257f1b05e520c0c5352ccba875d995aeb74c40c/tests/test_model.py#L667

https://github.com/ShopRunner/collie_recs/blob/6257f1b05e520c0c5352ccba875d995aeb74c40c/tests/test_model.py#L876

https://github.com/ShopRunner/collie_recs/blob/6257f1b05e520c0c5352ccba875d995aeb74c40c/tests/test_model.py#L894

Is that what you had in mind? If not, I can add an additional set of tests to target this better. Let me know!

[the remark this answers is not in the document]
Comment (peterskipper) [reply collie_recs/interactions/dataloaders.py:80]: I think that covers most of it! If the agreed upon formats for the data loader come in 3 flavors:
> ((X, Y), Z) = ((user IDs, item IDs), negative item IDs) = Implicit Model
> (X, Y, Z) = (user IDs, item IDs, ratings) = Explicit Model
> ((X, Y), Z, C) = ((user IDs, item IDs), negative item IDs, interaction type) = Mash Up Implicit Model

Then we should have tests for both explicit AND implicit (I def saw the explicit one, is there sthg similar for implicit? Not sure how the hdf5 tests play into this). If we are making changes to the data loader to allow for Mashup now, then we should add a test for that dataloader as well. If that comes later when I add my PR, we should obviously hold off.

[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/interactions/dataloaders.py:80]: I'll add for the implicit tests a check that ensures that's the right structure of the output. Then, I think we should be fully covered for the above structure (minus the eventual mashup).

**Review by peterskipper — APPROVED** [review]
LGTM!

**Comment by peterskipper** [comment]
https://media.giphy.com/media/jmS6YojdAaYw5z1LHi/giphy.gif
